Geo: Use more precise map markers in geo activities.

zestyping opened the pull request:
Issues: This feature is part of the [Geo roadmap](https://docs.google.com/document/d/1r3wJrOVDeFQkk4kWMoW29LtLI1vbBChqJ65w1Jyg4YU/edit#).
Risk: Low
Code scope: MapFragment implementations
UX scope: GeoPoint, GeoTrace, GeoShape
- [x] `./gradlew checkCode` passed
- [x] UI elements follow [style guidelines](http://bit.ly/2XtluID)

Requested reviewers: @cooperka, @grzesiek2010 

#### User-visible changes <!-- Required.  If no user-visible changes, write "None." -->
In GeoPoint, GeoTrace, and GeoShape, the entered point or points are drawn as small red circles, instead of big red or black balloons.  The smaller icon makes it easier to see precisely where the points are.

The user's location is displayed using a set of crosshairs at the location, as well as a semitransparent blue circle showing the 68% confidence radius of the location fix.

#### Notes for reviewers <!-- Where should reviewers start?  What explanations or diffs might help them understand your changes?  Omit for small, obvious PRs. -->

I recommend reviewing these commits one by one.

#### Notes for testers <!-- What should testers focus on?  What are potential regression risks?  Omit if UX scope is small and risks are minimal. -->

This is intended to be a 100% cosmetic change.  Only icons change; no behaviour should change.

#### Verification performed <!-- What has been done to ensure this works?  Omit for documentation-only PRs. -->

I have tried using GeoPoint, GeoTrace, and GeoShape, both with OSM and with Google Maps.

zestyping committed:
Geo: Use more precise map marker icons.

zestyping committed:
Geo: Use crossshairs to indicate location in GoogleMapFragment.

zestyping committed:
Geo: Fix up polygon drawing in GoogleMapFragment to match OsmMapFragment.

zestyping commented:
The location indicator becomes a set of blue crosshairs with a semitransparent blue circle showing the 68% confidence radius.  Collected points are now shown as small red circles.

Note that in OSM, the crosshairs are rotated according to the orientation of the user.  This is not ideal, but there is no way to tell the OSM API to not rotate the crosshairs.  With Google Maps, the crosshairs remain upright as intended.

With OSM:
![device-2019-03-08-043211](https://user-images.githubusercontent.com/236086/54029769-7ceaba80-415e-11e9-888c-842b9094c649.png)
![device-2019-03-08-043355](https://user-images.githubusercontent.com/236086/54029779-82480500-415e-11e9-80ec-b18b4e8fb21d.png)

With Google Maps:
![device-2019-03-08-043220](https://user-images.githubusercontent.com/236086/54029797-912eb780-415e-11e9-9157-b3a7313126ed.png)
![device-2019-03-08-044025](https://user-images.githubusercontent.com/236086/54029804-9855c580-415e-11e9-9bac-fcbae4f6e21e.png)

zestyping committed:
Fix Robolectric breakage of GoogleMap in tests.

cooperka commented on collect_app/src/main/java/org/odk/collect/android/map/GoogleMapFragment.java:
```diff
@@ -321,6 +326,33 @@ protected void moveOrAnimateCamera(CameraUpdate movement, boolean animate) {
+                .radius(radius)
+                .strokeWidth(1)
+                .strokeColor(0x800099ff) // matches the fillColor in ic_crosshairs.xml
+                .fillColor(0x200099ff)
```
Let's put these in `colors.xml`

zestyping replied:
Done.

cooperka commented on collect_app/src/main/res/drawable/ic_crosshairs.xml:
```diff
@@ -0,0 +1,9 @@
+        android:viewportWidth="48"
+        android:viewportHeight="48">
+    <path
+        android:fillColor="#FF0099FF"
```
also colors.xml

Is this color defined anywhere externally? Who came up with it? Might be nice to document.

zestyping replied:
I picked it to approximately match the blue that is used for the location dot and the shaded accuracy circle in Google Maps.  So, in some sense it's a guess.  It looks pretty close to my eye though.

zestyping replied:
Okay, I moved the colours into colors.xml and gave them all names.  Seemed like it wouldn't hurt to be thorough so I did them all.

cooperka commented on collect_app/src/main/res/drawable/ic_red_point.xml:
```diff
@@ -0,0 +1,8 @@
+        android:height="36dp"
+        android:viewportWidth="36"
+        android:viewportHeight="36">
+    <path android:fillColor="#ffff0000" android:pathData="M18,13c-2.77,0 -5,2.24 -5,5s2.24,5 5,5s5,-2.24 5,-5s-2.24,-5 -5,-5z"/>
```
(probably fine here since the file is explicitly called `red_point` and this is pure red)

zestyping replied:
I renamed this `map_point` because I moved the colour into `colors.xml`.

cooperka commented on collect_app/src/main/java/org/odk/collect/android/map/GoogleMapFragment.java:
```diff
@@ -493,6 +496,9 @@ protected Marker createMarker(GoogleMap map, MapPoint point, boolean draggable)
 
     protected BitmapDescriptor getBitmapDescriptor(int drawableId) {
+        if (testMode) {
+            return null; // during Robolectric tests, BitmapDescriptorFactory doesn't work
```
This is unfortunate. Do you think it would be worth stubbing out `BitmapDescriptorFactory` so you don't need to include test logic in the real class?

zestyping replied:
It turns out this guard actually isn't needed anymore—tests don't go through this code path because `map` is `null`.

It would be nice to be able to fix this with little stubs, but really the bigger issue is that we can't get `GoogleMap` to work at all under Robolectric.  Since the whole `map` is `null` there's a lot of other stuff that just has to be skipped in tests.  (Lots of people have complained about this—writing a stub for `GoogleMap` would be nice too but it's a bigger project.)

Anyway, I was at least able to take out these two lines, so there's less testing cruft in here.

cooperka commented on collect_app/src/main/java/org/odk/collect/android/utilities/IconUtils.java:
```diff
@@ -29,4 +34,15 @@ public static int getNotificationAppIcon() {
     }
+
+    /** Renders a Drawable (such as a vector drawable) into a Bitmap. */
+    public static Bitmap getBitmap(Context context, int drawableId) {
```
Since others may use this method in the future, can you add the optimizations from [this SO answer](https://stackoverflow.com/a/10600736/763231)? Overall very similar, but checks a few edge cases.

zestyping replied:
Yep!  Added handling for bitmaps and solid colours.

cooperka reviewed (commented):
Looks good, just some small suggestions.

zestyping committed:
Remove unnecessary testMode guard in getBitmapDescriptor method.

zestyping committed:
Move colour constants used in geo activities into named colour resources.

zestyping committed:
Make getBitmap properly handle different kinds of Drawables.

zestyping commented:
Thanks for the review!  I think I've addressed all the comments now.

zestyping committed:
Fix incorrect colour constants.  Fix clearing of features from GoogleMapFragment.

zestyping commented:
Oops.  Caught a bug with the colour constants—the functions take ints, and it's too bad there's no type distinction between ints that are resource IDs and ints that are colour values.

Also, tapping the trash button previously cleared the crosshairs off the map as well.  That's fixed now.

I think this is in good shape :)

cooperka reviewed (approved):
🥇

cooperka commented:
@opendatakit-bot label "needs testing"

mmarciniak90 commented:
@zestyping I noticed that dot is not in the middle of the pointer in **OSM**. 
This behavior is visible on **Android 8.1, 6.0, 5.1, 4.4**

| GeoPoint | GeoTrace | GeoShape|
|---|---|---|
|![Screenshot_20190321-151842](https://user-images.githubusercontent.com/3907554/54758669-e5b94480-4bec-11e9-978c-9064ed533db4.png)|![Screenshot_20190321-152148](https://user-images.githubusercontent.com/3907554/54758839-39c42900-4bed-11e9-9301-3757237d50f7.png)|![Screenshot_20190321-151817](https://user-images.githubusercontent.com/3907554/54758693-f5388d80-4bec-11e9-8f85-66947db85a21.png)|

zestyping commented:
@mmarciniak90 This is very puzzling!  I have tried this on three different phones and the circle always appears in the center of the crosshairs for me.  I'm going to keep trying more phones in the emulator.  If you have any further details about the phones you used or the conditions under which you saw this, that would be helpful!  Was the circle sometimes aligned and sometimes not aligned with the crosshairs, or was it always misaligned?

zestyping commented:
Interesting.  I see that in all three of your screenshots, the blue circle is centered on the midpoint of the downward segment of the crosshairs.

zestyping commented:
I've now tried three phones, a tablet, and two emulated phones, and still the blue circle and crosshairs appear to be correctly aligned on all of them.

Sony Android 6.0.1 phone
Samsung Android 6.0.1 phone
Huawei Android 5.1 phone
Samsung Android 8.1.0 tablet
Hi-res Android 6.0 phone in emulator (1440 x 2560)
Lo-res Android 6.0 phone in emulator (768 x 1280)

mmarciniak90 commented:
@zestyping yes, I agree, it is puzzling. 
I show you screenshots from both maps.

Android 8.1

| Google Maps | OSM|
|---|---|
|![Screenshot_20190325-151446](https://user-images.githubusercontent.com/3907554/54926893-83748280-4f11-11e9-9178-a47764c0b6e8.png)|![Screenshot_20190325-151521](https://user-images.githubusercontent.com/3907554/54926891-82435580-4f11-11e9-983d-2c052d45088d.png)|

zestyping commented:
@mmarciniak90 To investigate this problem, I need more information on how you encountered it.  Can you please tell me more about the exact specifications of the model or emulator you used, and the exact conditions under which it occurred or did not occur?  What was the screen resolution?  What steps did you take?  What would you recommend that I do to replicate exactly the situation in which you encountered the issue?  Thanks!

yanokwa commented:
I am able to reproduce this on an Moto G 5th Gen running Android 7.0 running this branch. I'm using the OSM SDK with CartoDB Positron with a geotrace question and no appearances. 

![offset-small](https://user-images.githubusercontent.com/32369/54953912-9766b700-4f06-11e9-88db-c2f2dd2d2933.png)

I cannot reproduce this on a Samsung Galaxy Tab A 2016 running Android 5.1.1.

zestyping commented:
Awesome, thanks @yanokwa!  Looks like there's something weird happening with the map labels there?

What's the resolution of your screen?

My latest hypothesis is that the bounding box for the SVG of the crosshairs is computed incorrectly such that the dimensions of the rasterized image are taller than they should be, so the image is centered incorrectly.  Just a guess.

yanokwa commented:
I put noise over the labels so my address isn't part of the public record.

The Moto G5 is 1920x1080. The testers use the devices listed at https://github.com/opendatakit/collect#contributing-testing.

@cooperka Tested on a Pixel 2 (1920×1080) with Android 9 and that worked.

grzesiek2010 commented:
I can see it even on the emulator (API 23 - I didn't test others):
![Screenshot_1553596480](https://user-images.githubusercontent.com/3276264/54990562-5086c900-4fbb-11e9-8074-40421768cb89.png)

@zestyping are you sure you use OSM maps? The problem is only there.

mmarciniak90 commented:
@opendatakit-bot unlabel "needs testing"

zestyping commented:
@grzesiek2010 Yes, absolutely sure.  All six devices I tested were fine while using the OSM basemap.  What emulator did you use, what system image, and what screen resolution?  Please tell me everything you can so I can try to duplicate your conditions.

@mmarciniak90 Could you please tell me the answers to the questions in my last comment to you?  Did the problem appear on all the devices you tested, or only some? I really need as much information as possible to figure out why this is happening.

grzesiek2010 commented:
For example:
![Screenshot from 2019-03-26 15-26-52](https://user-images.githubusercontent.com/3276264/55005647-acae1500-4fdc-11e9-92ab-fc7d932b7a9b.png)
but I tried two others with the same result, so that's really strange you can't reproduce it.

mmarciniak90 commented:
@zestyping I see this problem on devices:
* Samsung Galaxy J1 SM-J100H - Android 4.4.4
* Huawei Y560-L01 - Android 5.1.1
* Sony Xperia Z3 D6603 - Android 6.0.1
* Samsung Galaxy S7 SM-G930F - Android 7.0.0
* LG Nexus 5X - Android 8.1

I used 'All widgets' form.

cooperka commented:
I do **not** experience this bug with either _Pixel 2 API 28_ or _Galaxy Nexus API 23_ emulated (very similar to @grzesiek2010's device which does experience this). Everything looks fine, definitely using OSM, tried every type of geo activity and they're all fine for me.

![Screenshot_1553614612](https://user-images.githubusercontent.com/2047062/55010914-790ec300-4fbb-11e9-8f6c-1ce009b2866a.png)

Resolutions:

<img width="750" alt="Screen Shot 2019-03-26 at 11 40 01" src="https://user-images.githubusercontent.com/2047062/55011174-ecb0d000-4fbb-11e9-97c9-3076c56a0453.png">

grzesiek2010 committed:
Fixed the arrow position in OSM maps

The default marker used by OSM is a person. The location of the person marker's feet is adjusted based on display density. Since we are using crosshairs which are square instead of the person marker, the marker location needs to be adjusted.

grzesiek2010 commented:
This commit seems to fix the problem https://github.com/grzesiek2010/collect/commit/eded9160a0ab682b572eba0b5d612016b1cacb01

Looks as if we need to setPersonHotspot just after setting the used bitmap (probably later it's adjusted automatically).

grzesiek2010 commented:
I tested on a real device and using the emulator.

lognaturel commented:
👏 👏 👏 How did you figure that out, @grzesiek2010?

grzesiek2010 commented:
>How did you figure that out, @grzesiek2010?

I investigated sourcecode https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/overlay/mylocation/MyLocationNewOverlay.java

in the constructor after calling [setDirectionArrow](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/overlay/mylocation/MyLocationNewOverlay.java#L109) the value of [mPersonHotspot](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/overlay/mylocation/MyLocationNewOverlay.java#L113) is set but using original resources

then in our code, we call [setDirectionArrow](https://github.com/osmdroid/osmdroid/blob/master/osmdroid-android/src/main/java/org/osmdroid/views/overlay/mylocation/MyLocationNewOverlay.java#L124)

but we don't adjust `mPersonHotspot`.

lognaturel commented:
That makes a lot of sense. @zestyping, that seems to close the case. Will you be able to make the change with a comment today? Then it should be possible to get it verified and hopefully merged before the code freeze for 1.21.

yanokwa commented:
I can confirm that this works on my Moto G5. 

@lognaturel, since we are heading towards a code freeze, can you push the fix into Ping's branch and mark this PR as needs testing?

lognaturel commented:
Hopefully this can get verified and merged during your day, @grzesiek2010. If for some reason there are other fixes that need to be done, I believe you can also push to @zestyping's branch. Alternately, you can add commits to your local branch and make a PR for that. 👍

mmarciniak90 commented:
**Tested with success**

Verified on Android: 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

Verified all GeoWidgets in `All widgets` form
Verified OSM and Google Maps.

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

zestyping commented:
Fantastic, @grzesiek2010!  Thanks for jumping in!  I came down with a nasty cold yesterday morning and have been out of commission trying to sleep it off.  Glad you found the bug!
